plugin/src/main/java/org/opensearch/ml/task/MLTaskManager.java

```diff
@@ -333,30 +337,33 @@ public void updateMLTask(
             } catch (InterruptedException e) {
                 log.error("Failed to acquire semaphore for ML task {}", taskId, e);
                 listener.onFailure(e);
-                return; // return directly if can't get semaphore
+                return; // return directly if you can't get semaphore
             }
             try {
                 if (updatedFields == null || updatedFields.isEmpty()) {
                     listener.onFailure(new IllegalArgumentException("Updated fields is null or empty"));
                     return;
                 }
                 UpdateRequest updateRequest = new UpdateRequest(ML_TASK_INDEX, taskId);
-                Map<String, Object> updatedContent = new HashMap<>();
-                updatedContent.putAll(updatedFields);
+                Map<String, Object> updatedContent = new HashMap<>(updatedFields);
                 updatedContent.put(LAST_UPDATE_TIME_FIELD, Instant.now().toEpochMilli());
                 updateRequest.doc(updatedContent);
-                updateRequest.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
-                if (updatedFields.containsKey(STATE_FIELD) && TASK_DONE_STATES.contains(updatedFields.containsKey(STATE_FIELD))) {
-                    updateRequest.retryOnConflict(3);
-                }
-                ActionListener<UpdateResponse> actionListener = semaphore == null
-                    ? listener
-                    : ActionListener.runAfter(listener, semaphore::release);
-                try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
-                    client.update(updateRequest, ActionListener.runBefore(actionListener, context::restore));
-                } catch (Exception e) {
-                    actionListener.onFailure(e);
-                }
+
+                UpdateDataObjectRequest updateDataObjectRequest = UpdateDataObjectRequest
+                    .builder()
+                    .index(ML_TASK_INDEX)
+                    .id(taskId)
+                    .tenantId(tenantId)
+                    .dataObject(updatedContent)
+                    .retryOnConflict(3)
+                    .build();
+
+                sdkClient
+                    .updateDataObjectAsync(updateDataObjectRequest, client.threadPool().executor(GENERAL_THREAD_POOL))
+                    .whenComplete((r, throwable) -> {
+                        semaphore.release();
```

Comment:
Previous code null-checks this semaphore. Is that needed here?

Reply:
Yeah I added it back. Thanks.